pdns/pdns_recursor.cc

```diff
@@ -2566,21 +2658,80 @@ static void handleRunningTCPQuestion(int fd, FDMultiplexer::funcparam_t& var)
         return;
       }
       else {
+        // We have read a proper query
         ++g_stats.qcounter;
         ++g_stats.tcpqcounter;
-        ++conn->d_requestsInFlight;
-        if (conn->d_requestsInFlight >= TCPConnection::s_maxInFlight) {
-          t_fdm->removeReadFD(fd); // should no longer awake ourselves when there is data to read
+
+        string response;
+        uint32_t qhash = 0;
+        RecursorPacketCache::OptPBData pbData{boost::none};
+
+        /* It might seem like a good idea to skip the packet cache lookup if we know that the answer is not cacheable,
+           but it means that the hash would not be computed. If some script decides at a later time to mark back the answer
+           as cacheable we would cache it with a wrong tag, so better safe than sorry. */
+        bool cacheHit = checkForCacheHit(qnameParsed, dc->d_tag, conn->data, qname, qtype, qclass, g_now, response, qhash, pbData, true, dc->d_source);
+        dc->d_qhash = qhash;
+
+        if (cacheHit) {
+          if (t_protobufServers && dc->d_logResponse && !(luaconfsLocal->protobufExportConfig.taggedOnly && pbData && !pbData->d_tagged)) {
+            struct timeval tv{0, 0};
+            protobufLogResponse(dh, pbData, tv, true, dc->d_source, dc->d_destination, dc->d_ednssubnet, dc->d_uuid, dc->d_requestorId, dc->d_deviceId, dc->d_deviceName);
+          }
+
+          if (!g_quiet) {
+            g_log<<Logger::Notice<<t_id<< " TCP question answered from packet cache tag="<<dc->d_tag<<" from "<<dc->d_source.toStringWithPort()<<(dc->d_source != dc->d_remote ? " (via "+dc->d_remote.toStringWithPort()+")" : "")<<endl;
+          }
+
+          bool hadError = sendResponseOverTCP(dc, response);
```

Comment:
I wonder if we could try to factorize that code, which seems duplicated from `startDoResolve`?

Reply:
Yes, it's similar, though a bit more simpler, lacking the `requestsInFlight` accounting since it can answer immediately. I will take a look.